### PR TITLE
[AMDGPU] Simplify API of matchFPExtFromF16. NFC.

### PR DESCRIPTION
**Pull request by jayfoad**

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp [hunk 1]
```diff
@@ -353,23 +353,20 @@ bool GCNTTIImpl::canSimplifyLegacyMulToMul(const Instruction &I,
...
-      return false;
-
-    FPExtSrc = ConstantFP::get(Type::getHalfTy(Arg->getContext()), Val);
-    return true;
+    if (!LosesInfo)
+      return ConstantFP::get(Type::getHalfTy(Arg->getContext()), Val);
   }
-
-  return false;
+  return {};
```
Comment (shiltian) [inline llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp:369]:
```suggestion
  return std::nullopt;
```

File: llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp [hunk 2]
```diff
@@ -353,23 +353,20 @@ bool GCNTTIImpl::canSimplifyLegacyMulToMul(const Instruction &I,
 }
 
 /// Match an fpext from half to float, or a constant we can convert.
-static bool matchFPExtFromF16(Value *Arg, Value *&FPExtSrc) {
-  if (match(Arg, m_OneUse(m_FPExt(m_Value(FPExtSrc)))))
-    return FPExtSrc->getType()->isHalfTy();
-
+static std::optional<Value *> matchFPExtFromF16(Value *Arg) {
+  Value *Src;
```
Comment (shiltian) [inline llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp:357]: I understand here you just follow existing code, but I'd just initialize them to `nullptr`.

**Review by shiltian — APPROVED** [review]
Not sure what the real benefit is but it's fine.

**Comment by jayfoad** [comment]
> Not sure what the real benefit is

I find it much cleaner to return things via the return value of a function, instead of modifying one of its arguments.

File: llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp (hunk 1 above)
Comment (jayfoad) [reply llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp]: Done

File: llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp (hunk 2 above)
Comment (jayfoad) [reply llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp]: Done

**Review by shiltian — APPROVED** [review]

**Comment by jayfoad** [comment]
> > Not sure what the real benefit is
> 
> I find it much cleaner to return things via the return value of a function, instead of modifying one of its arguments.

I'd also be happy with returning `nullptr` on failure, instead of using `optional`.

**Comment by shiltian** [comment]
Yeah, it seems like the boolean return value has the same semantics as `nullptr`.,

**Review by shiltian — APPROVED** [review]
LGTM. Thanks.

**Pull request merged**